### PR TITLE
Fix NativePin for Pi 4

### PR DESCRIPTION
**Pull request by waveform80**
This is a re-factor and re-base of @lurch's #900 PR. The initial commit is from the original PR but re-based onto the current master branch, and I've done a bit of re-factoring work in the second commit which makes the d-t querying a bit more generic and splits the pull-handling in two refinements of the NativePin class.

### CODE REVIEW

**Comment by waveform80** [comment]
Tested under Python 2.7 as well, locally on a Pi 4. I've also tested the implementation on an old Pi 1, Pi 2, Pi 3B+, and a Pi 4.

**Comment by lurch** [comment]
I'm happy to give this the thumbs-up if you've verified it as working properly on various different models of RPi. :+1: 

Will be good to see this finally merged, so long after the release of the Pi4 :grinning:

**Comment by waveform80** [comment]
> I'm happy to give this the thumbs-up if you've verified it as working properly on various different models of RPi. +1
> 
> Will be good to see this finally merged, so long after the release of the Pi4 grinning

Absolutely, and yes please do - I've verified it under current RaspiOS on a Pi1 (2835), Pi2 (2836), Pi3+ (2837), and Pi4 (2711). Haven't given it a whirl yet on a CM4/400 but I did verify device-tree parsing code was being correctly used in all those cases rather than the dict-fallback so I'd be reasonably confident that even if the CM4/400 *did* have a difference there (which I'm sure they don't) it shouldn't matter provided the d-t reflected it accurately.

If you've got any spare time, you could cast your expert eye over the RotaryEncoder PR? (I've been noticing I'd missed stupid things like the bounce-time initializers!) Not to worry if not though - it's had a bit of third-party testing already so I'm reasonably confident in it. Just having a crack at getting something for 7-seg displays in too (singletons only). Basically I'm hoping we can cut a release this week, as I need to get a version of gpiozero which includes lgpio into Ubuntu 21.04 and we're already into the freeze period.

**Comment by lurch** [comment]
> even if the CM4/400 _did_ have a difference there (which I'm sure they don't)

The source-code for the DeviceTree stuff is all at https://github.com/raspberrypi/linux/tree/rpi-5.10.y/arch/arm/boot/dts if you wanted to investigate the differences between different models.

> If you've got any spare time, you could cast your expert eye over the RotaryEncoder PR?

Normally I would, but I'm afraid I've been quite busy with https://github.com/raspberrypi/pico-sdk/ recently....

It's been a _long_ time since I've looked at GpioZero, so I suspect it'd take me quite a while to get up to speed again.

File: gpiozero/pins/native.py [hunk 1]
```diff
@@ -324,7 +425,10 @@ def __init__(self):
         self.mem = GPIOMemory(self.pi_info.soc)
         self.fs = GPIOFS(self, queue)
         self.dispatch = NativeDispatchThread(self, queue)
-        self.pin_class = NativePin
+        if self.pi_info.soc == 'BCM2711':
+            self.pin_class = Native2711Pin
+        else:
+            self.pin_class = Native2835Pin
```
Comment (lurch) [inline gpiozero/pins/native.py:431]: So I guess this means that `NativePin` itself will never be directly used? Would it be worth putting in an exception or something to enforce that?

**Review by lurch — APPROVED** [review]
I've not tested this, but it looks similar enough to my previous PR, which I did test.

File: gpiozero/pins/native.py (hunk 1 above)
Comment (waveform80) [reply gpiozero/pins/native.py:431]: Yes, NativePin is effectively an abstract class at this point. I've never been hugely fussed about banning the instantiation of abstract classes; it's one of those issues that's come up sufficiently rarely and (provided abstract methods are marked with some appropriate error like NotImplementedError) is pretty obvious if/when it does that it's not a headache to track down. In this case I think I'll leave it - I don't particularly want to complicate the hierarchy at the pin level with Python's abstractclass machinery.

**Comment by waveform80** [comment]
> > even if the CM4/400 _did_ have a difference there (which I'm sure they don't)
> 
> The source-code for the DeviceTree stuff is all at https://github.com/raspberrypi/linux/tree/rpi-5.10.y/arch/arm/boot/dts if you wanted to investigate the differences between different models.

Thanks for the link! I've dug through diffs of the d-t's between Ubuntu and RaspiOS before (usually when debugging u-boot stuff) but haven't had cause to dig into the GPIO related nodes before. I'll take a look a bit later! 

> > If you've got any spare time, you could cast your expert eye over the RotaryEncoder PR?
> 
> Normally I would, but I'm afraid I've been quite busy with https://github.com/raspberrypi/pico-sdk/ recently....
> 
> It's been a _long_ time since I've looked at GpioZero, so I suspect it'd take me quite a while to get up to speed again.

Fair enough :) The Pico SDK's docs are looking very professional (been skimming the C PDF ones, though I've yet to find time to actually *use* it yet)! I should take a look at asciidoc at some point...

**Comment by waveform80** [comment]
> I've not tested this, but it looks similar enough to my previous PR, which I did test.

Yup, same idea basically: query the d-t for the GPIO registers, map the address to the CPU's view, fallback to the model lookup in a dict if that fails for I/O related reasons. I'll get @bennuttall to test it on some hardware later (so it's more than just me that's run this :) then we can merge it.

File: gpiozero/pins/native.py (hunk 1 above)
Comment (lurch) [reply gpiozero/pins/native.py:431]: Should `_get_pull` and `_set_pull` "base methods" in `NativePin` raise `NotImplementedError` ?

**Comment by lurch** [comment]
> The Pico SDK's docs are looking very professional (been skimming the C PDF ones, though I've yet to find time to actually use it yet)!

Thanks :grin: Getting the Doxygen included into the AsciiDoc was "challenging" :wink: Found three different projects that claimed to be able to do that, but none of them turned out to be suitable for our purposes, so I had to do it via XSLT, which was a bit of a steep learning curve (99% of XSLT tutorials are geared at doing simple XML->HTML translations).

File: gpiozero/pins/native.py (hunk 1 above)
Comment (waveform80) [reply gpiozero/pins/native.py:431]: `_set_pull` effectively does by inheritance (admittedly not NotImplementedError, I think it's PinFixedPull, but still that'd be a pretty big hint to anyone running into it that an override for `_set_pull` was missing. But yes, `_get_pull` is missing one (it'll just return 'floating' by default). Yup, you've convinced me: I should chuck a `NotImplementedError` into those (won't cost anything at runtime as the super-class should never get called for those). Will do that in a mo.

**Comment by waveform80** [comment]
> > The Pico SDK's docs are looking very professional (been skimming the C PDF ones, though I've yet to find time to actually use it yet)! I should take a look at asciidoc at some point...
> 
> Thanks grin Getting the Doxygen included into the AsciiDoc was "challenging" wink Found three different projects that claimed to be able to do that, but none of them turned out to be suitable for our purposes, so I had to do it via XSLT, which was a bit of a steep learning curve... (99% of XSLT tutorials are geared at doing simple XML->HTML translations).

Ooohh, I'm actually a bit of a fan of XSLT - the new version of structa I've been working on uses it in possibly the stupidest way I've tried yet (as an interface rendering layer for the Linux terminal :). I really must get that finished at some point... And yes, I remember it being quite horrific to learn. I think I wound up just reading the spec and experimenting until a light-bulb went on it my head!

**Review by lurch — APPROVED** [review]

**Comment by waveform80** [comment]
Well, getting this working on Ubuntu (and thus arm64 too) turned out to be more interesting than I'd expected, but it's now working and passing the test suite. @bennuttall this is ready for merge :)

File: gpiozero/pins/native.py [hunk 2]
```diff
@@ -205,6 +205,17 @@ def __init__(self, soc):
...
         self.mem = mmap.mmap(self.fd, 4096, offset=offset)
+        # Register reads and writes must be in native format (otherwise
+        # struct resorts to individual byte reads/writes and you can't hit
+        # half a register :). For arm64 compat we have to figure out what the
+        # native unsigned 32-bit type is...
+        try:
+            self.reg_fmt = {
+                struct.calcsize(fmt): fmt
+                for fmt in (nstr('@I'), nstr('@L'))
+            }[4]
```
Comment (lurch) [inline gpiozero/pins/native.py:216]: Oh boy... more impenetrable @waveform80 code :wink:

**Comment by lurch** [comment]
Whoop! :tada:

**Pull request merged**